test/distributed/test_distributed.py

```diff
@@ -847,6 +847,43 @@ def test_reduce_full_group_max(self):
         group, group_id, rank = self._init_full_group_test()
         self._test_reduce_helper(group, group_id, rank, dist.ReduceOp.MAX, -1, 10, 10)
 
+    @skip_if_no_gpu
+    @require_backend({"gloo", "nccl"})
+    def test_all_reduce_result_cuda(self):
+        group, group_id, rank = self._init_global_test()
+        rank_to_GPU = self._init_multigpu_helper()
+        for src in group:
+            if rank == src:
+                tensor = _build_tensor(src + 1, 2)
+            else:
+                tensor = _build_tensor(src + 1, 10)
+            tensor = tensor.cuda(rank_to_GPU[rank][0])
+
+            opts = AllreduceOptions()
+            opts.reduceOp = dist.ReduceOp.SUM
+
+            if group_id == GroupMember.WORLD:
+                work = _get_default_group().allreduce([tensor], opts)
+            else:
+                work = group_id.allreduce([tensor], opts)
+
+            # Calling result right the work is finished should throw exception.
```

Comment:
Is there a better way to guarantee under some scenario that this async work is not completed after calling into the API? (maybe @mrshenli may have some suggestions). If not, then this try-except approach looks like the best we can do.

Reply:
Not sure. In the case of Gloo, almost always we run into exception, but for NCCL, it is too fast, and very likely that the exception does not happend.